pkg/controller/plan/adapter/vsphere/builder.go

```diff
@@ -1176,32 +1184,223 @@ func (r *Builder) LunPersistentVolumeClaims(vmRef ref.Ref) (pvcs []core.Persiste
 	return
 }
 
+// FIXME rgolan - the behaviour needs to be per disk hense this method is flawed. Needs a bigger change.
+// For now this method returns true, if there's a mapping (backend by copy-offload-mapping ConfigMap, that
+// maps StoragetClasses to Vsphere data stores
 func (r *Builder) SupportsVolumePopulators() bool {
+	if !settings.Settings.Features.CopyOffload {
+		return false
+	}
+	dsMapIn := r.Context.Map.Storage.Spec.Map
+	for _, m := range dsMapIn {
+		ref := m.Source
+		ds := &model.Datastore{}
+		err := r.Source.Inventory.Find(ds, ref)
+		if err != nil {
+			klog.Errorf("failed to get datastore to detect volume populators support: %s", err)
+			return false
+		}
+
+		if m.OffloadPlugin != nil && m.OffloadPlugin.VSphereXcopyPluginConfig != nil {
+			klog.V(2).Infof("found offload plugin: config %+v on ds map  %+v", m.OffloadPlugin.VSphereXcopyPluginConfig, dsMapIn)
+			return true
+
+		}
+	}
 	return false
 }
 
+// PopulatorVolumes creates PVC in case the their are needed for the disks
+// in context, and according to the offload plugin configuration in the StorageMap
 func (r *Builder) PopulatorVolumes(vmRef ref.Ref, annotations map[string]string, secretName string) (pvcs []*core.PersistentVolumeClaim, err error) {
-	err = planbase.VolumePopulatorNotSupportedError
-	return
+	vm := &model.VM{}
+	err = r.Source.Inventory.Find(vm, vmRef)
+	if err != nil {
+		err = liberr.Wrap(err, "vm", vmRef.String())
+		return
+	}
+
+	dsMapIn := r.Context.Map.Storage.Spec.Map
+	for i := range dsMapIn {
+		mapped := &dsMapIn[i]
+		ref := mapped.Source
+		ds := &model.Datastore{}
+		fErr := r.Source.Inventory.Find(ds, ref)
+		if fErr != nil {
+			err = fErr
+			return
+		}
+
+		pvblock := core.PersistentVolumeBlock
+		for _, disk := range vm.Disks {
+			if disk.Datastore.ID == ds.ID {
+				storageClass := mapped.Destination.StorageClass
+				coldLocal, vErr := r.Context.Plan.VSphereColdLocal()
```

Comment:
The `VSphereColdLocal` does not exist anymore, it was renamed to `ShouldUseV2vForTransfer`, but I don't think that we need to check it here.

Reply:
ack

Reply:
In migration.go, there is a part which check SupportsVolumePopulator, and this returns `true` if the storagemap matches. I don't think that code is cold/warm aware.

Reply:
The name of the method was never perfect, it determines if the virt-v2v should do the transfer of the disks for us. As we are using volume populators the result from it should be false as we are managing the transfer ourselves. 
You could add some logic to the method to include also check for volume populators

Reply:
since populator volumes are exclusively needed in cold for vsphere, would that be clearer to just return with early with no pvcs, in case plan.spec.warm?
```
func (r *Builder) PopulatorVolumes(....) {

    if r.Plan.Spec.Warm {
      return fmt.Errorf("warm migration and populator volumes dont mix!")
      }
      ...
 }
 ```